### ISSUE TITLE
How do things like yarn installs work within nodenv?

### ISSUE THREAD

**Issue by vindard** [nodenv/nodenv#177]
**Context:** _I don't have a system version of NodeJS installed and I use node exclusively through `nodenv`._

I'd like to use yarn but when I run `$ npm i -g yarn`, after it completes I still don't have access to a `$ yarn` command in my terminal. I do have access to `$ npx yarn`, but I'm not sure if that's ideal or at what level yarn caching would work etc. in this scenario.

**Questions:**
- how do I get `yarn` to be a command at the global level?
- what happens if I uninstall a given version of npm? (do all yarn-installed modules go away from that version or is there some global thing outside of nodenv)?

**Comment by vindard** [comment]
Closing this issue, it looks like things resolved themselves once I rebooted my system.

For context for anyone else struggling with this, the binaries for global installs get installed and then aliased in the `$(nodenv root)/versions/<version>/bin/` directory where they're then accessible via the `$PATH`.

e.g. for node v16.1.0 my binaries were in `~/.nodenv/versions/16.1.0/bin`

**Comment by jasonkarns** [comment]
@vindard Glad you got it sorted.

**TL;DR** Mostly likely, you just needed to run `nodenv rehash` after installing yarn. That wouldn't require a system restart (nor does it require a shell restart).

In more detail:

Node's default location for global modules is _within_ its own installed directory. So node 16.1.0 would install its global modules to `$(nodenv root)/versions/16.1.0/bin`. (Notice to homebrew users: homebrew _customizes_ this location such that global modules are preserved across node upgrades; hence why the behavior is different between a homebrew-installed node and one installed using a version manager.)

nodenv then creates shims for all executables found in all of your nodes' `bin` directories. These shims are added to your path by the `nodenv init` line that is added to your shell init script. (run `nodenv shims` to see them all)

These shims are what get executed when you run a "global" command like `npm` or `yarn` or even `node` itself. The shim is what is responsible for invoking nodenv and determining which of your node versions to "delegate" the command to.

What most likely happened to cause you trouble: you didn't have a yarn shim created after installing yarn, so the `yarn` executable wasn't in your PATH.

Something has to trigger nodenv to tell it when a new global module is installed, so that nodenv can create a shim for the executable(s). When you install a new node, there are callback hooks that run to make shims for all the executables in `bin/`; but when you later install global modules, nothing has told nodenv to create a shim for the yarn executable. In short: after installing a node module globally: run `nodenv rehash`. That will create shims for all the executables it finds in bin (including any newly installed global modules.

There is a plugin that used to make this unnecessary: https://github.com/nodenv/nodenv-package-rehash. It worked by running `nodenv-rehash` automatically using npm's global "postinstall" hooks. Unfortunately, npm has deprecated those hooks with npm7, so I don't believe that plugin works any more. It's also worth noting that now the yarn shim exists, you won't need to run `nodenv rehash` after installing yarn in any _other_ nodes; but it's still good practice to run rehash after every global module install/uninstall.

Also, FWIW, there's another nodenv plugin https://github.com/nodenv/nodenv-default-packages which allows you to define a list of packages to install (globally) in every node. `yarn` is a prime candidate for that list. This way, whenever you install a new version of node, the default-packages plugin will ensure yarn is installed (globally) in that new node.

**Comment by vindard** [comment]
Thanks @jasonkarns, this all really great info!

I did install the rehash plugin as well, but I'll have to keep in mind that it may not always automatically work. Good to know that that's what `rehash` does 👍🏽
